### PR TITLE
8261730: C2 compilation fails with assert(store->find_edge(load) != -1) failed: missing precedence edge

### PR DESCRIPTION
**Pull request by robcasloz**
This change relaxes an assertion in `PhaseCFG::verify()` to accept the case where a store is used to implement an implicit null check and a load is placed in the null block.

The implicit null check transformation `PhaseCFG::implicit_null_check()` tries to reuse globally scheduled memory operations to perform null checks, hoisting the memory operations into null-check blocks if necessary. When hoisting stores, `implicit_null_check()` preserves anti-dependences between the hoisted store and loads in the corresponding null-check and not-null blocks, but disregards possible anti-dependences with loads in the null block. This is safe, as the null block is only executed when the hoisted store has "failed" (attempted to write to the null address). This change makes `PhaseCFG::verify()` aware of this fact to avoid falsely reporting an anti-dependence violation (missing precedence edge from the load in the null block to the store hoisted to the null-check block).

**Example.** After global scheduling ([before-implicit-null-check.pdf](https://github.com/openjdk/jdk/files/6060581/before-implicit-null-check.pdf)), `implicit_null_check()` detects a store `(17 storeImmIO)` in B6 that could potentially perform the null check done by `(5 testN_reg)` in B4. After checking that neither B4 or B6 contain anti-dependent loads, the transformation replaces `(5 testN_reg)` with `(17 storeImmIO)` by hoisting the latter into B4 and adding an additional `(46 NullCheck)` pseudo-operation ([after-implicit-null-check.pdf](https://github.com/openjdk/jdk/files/6060614/after-implicit-null-check.pdf)). Finally, after local scheduling, `PhaseCFG::verify()` falsely reports that `(17 storeImmIO)` can overwrite the memory value read by the load operation `(34 subI_rReg_mem)` in the null block B5. This overwrite cannot happen, because executing B5 implies `(17 storeImmIO)`'s failure to store any value.

Tested on hs-tier1-5 (windows-x64, linux-x64, linux-aarch64, and macosx-x64).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261730](https://bugs.openjdk.java.net/browse/JDK-8261730): C2 compilation fails with assert(store->find_edge(load) != -1) failed: missing precedence edge

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2777/head:pull/2777`
`$ git checkout pull/2777`

### CODE REVIEW

File: src/hotspot/share/opto/gcm.cpp [hunk 1]
```diff
@@ -770,7 +770,24 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
...
+        Block* null_block = NULL;
+        if (store_null_check != NULL) {
+          Node* if_true = store_null_check->find_out_with(Op_IfTrue);
+          assert(if_true != NULL, "null check without null projection");
+          Node* null_block_region = if_true->find_out_with(Op_Region);
+          assert(null_block_region != NULL, "null check without null region");
+          null_block = get_block_for_node(null_block_region);
+        }
+#endif
+        assert((store_null_check != NULL && LCA == null_block) ||
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/gcm.cpp:789]: I think you could avoid the `store_null_check != NULL` check in the assert because `LCA` is always non-null.

**Review by TobiHartmann — APPROVED** [review]
Nice analysis. Your fix looks good to me (added a minor comment).

File: src/hotspot/share/opto/gcm.cpp (hunk 1 above)
Comment (robcasloz) [reply src/hotspot/share/opto/gcm.cpp]: Done, thanks!

**Comment by robcasloz** [comment]
> Nice analysis. Your fix looks good to me (added a minor comment).

Thanks for reviewing, Tobias! I just addressed your comment.

**Review by TobiHartmann — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]

**Comment by robcasloz** [comment]
Thanks for reviewing, Vladimir!

**Comment by robcasloz** [comment]
/summary
Relax assertion in PhaseCFG::verify() to accept the case where a store is used
to implement an implicit null check and a load is placed in the null block.

**Pull request closed without merging**